app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt

```diff
@@ -614,6 +617,14 @@ class BrowserTabFragment : Fragment(), FindListener, CoroutineScope {
         rootView.addOnLayoutChangeListener(logoHidingListener)
     }
 
+    private fun userSelectedAutocomplete(suggestion: AutoCompleteSuggestion) {
+        // send pixel before submitting the query and changing the autocomplete state to empty; otherwise will send the wrong params
```

Comment:
Use of `launch` first here still doesn't guarantee it'll happen before `onUserSubmittedQuery`. `launch` will happen "very soon but not immediately" and while sometimes it will run before the call to `onUserSubmittedQuery`, it isn't guaranteed. And it's probably pot luck what happens each time the code runs. See the screenshot below for proof... first time it was the order you expect, and the second time it's flipped the order.

For background info why this happens, see this post, and in particular the first example which is identical to the code you have: https://medium.com/@trionkidnapper/launching-a-kotlin-coroutine-for-immediate-execution-on-the-main-thread-8555e701163b 

I think we need a way to get a snapshot of the required data immediately, rather than waiting for the `fireAutocompletePixel` to run and retrieve the data itself. ie, i'm wondering if `fireAutocompletePixel` should perhaps take the values as params, so they can be constructed immediately and at the right time, rather than obtaining the values itself.

What's your thoughts on this?
<img width="672" alt="Screenshot 2019-07-24 at 11 33 38" src="https://user-images.githubusercontent.com/1336281/61787101-fc73e000-ae06-11e9-857d-1b6b57f0eb8a.png">

Reply:
That's good to know! I did think on passing the data to `fireAutocompletePixel` like your suggestion but moving around the calls seemed easier. I'll have a look at how that can be done. Thanks for letting me know :)

Reply:
This has now changed, `onUserSubmittedQuery` has been moved inside the GlobalScope so it fires the pixel first. I also made sure that if firing the pixel takes a long time it doesn't affect the call to `onUserSubmittedQuery`.